src/target/sam4l.c

```diff
@@ -0,0 +1,410 @@
+{
+	uint32_t cmd_reg;
+	uint32_t status;
+	int	timeout;
+	DEBUG("\nSAM4L: sam4l_flash_command: FSR: 0x%08x, page = %d, command = %d\n", 
+		(unsigned int)(FLASHCALW_FSR), (int) page, (int) cmd);
+	/* wait for Flash controller ready */
+	for (timeout = 0; timeout < FLASH_TIMEOUT; timeout++) {
+		status = target_mem_read32(t, FLASHCALW_FSR);
+		if (status & FLASHCALW_FSR_FRDY) {
+			break;
+		}
+	}
+	if (timeout == FLASH_TIMEOUT) {
+		DEBUG("\nSAM4L: sam4l_flash_command: Not ready! Status = 0x%08x\n", (unsigned int) status);
+		return -1; /* Failed */
+	}
+	/* load up the new command */
+	cmd_reg = (cmd & FLASHCALW_FCMD_CMD_MASK) |
+			  ((page & FLASHCALW_FCMD_PAGEN_MASK) << FLASHCALW_FCMD_PAGEN_SHIFT) |
+		  	  (0xA5 << FLASHCALW_FCMD_KEY_SHIFT);
+	DEBUG("\nSAM4L: sam4l_flash_command: Wrting command word 0x%08x\n", (unsigned int) cmd_reg);
+	/* and kick it off */
+	target_mem_write32(t, FLASHCALW_FCMD, cmd_reg);
+	/* don't actually wait for it to finish, the next command will stall if it is not done */
+	return 0;
+}
+
+/*
+ * Write data from 'src' into flash using the algorithim provided by
+ * Atmel in their data sheet.
+ */
+static int
+sam4l_flash_write_buf(struct target_flash *f, target_addr addr, const void *src, size_t len)
+{
+	target *t = f->t;
+	uint32_t *src_data = (uint32_t *)src;
+	uint32_t ndx;
+	uint16_t page;
+	
+	DEBUG("\nSAM4L: sam4l_flash_write_buf: addr = 0x%08lx, len %d\n", (long unsigned int) addr, (int) len);
+	/* This will fail with unaligned writes, the write_buf version */
+	while (len) {
+		page = addr / SAM4L_PAGE_SIZE;
+
+		/* clear the page buffer */
+		if (sam4l_flash_command(t, 0, FLASH_CMD_CPB)) {
+			return -1;
+		}
+
+		/* Now fill page buffer with our 512 bytes of data */
+
+		/* I did try to use target_mem_write however that resulted in the
+		 * last 64 bits (8 bytes) to be incorrect on even pages (0, 2, 4, ...)
+		 * since it works this way I've not investigated further.
+		 */
+		for (ndx = 0; ndx < SAM4L_PAGE_SIZE; ndx += 4) {
+			/*
+ 			 * the page buffer overlaps flash, its only 512 bytes long
+			 * and no matter where you write it from it goes to the page
+			 * you point it to. So we don't need the specific address here
+			 * instead we just write 0 - pagelen (512) and that fills our
+			 * buffer correctly.
+			 */
+			target_mem_write32(t, addr+ndx, *src_data);
+			src_data++;
+		}
+		/* write the page */
+		if (sam4l_flash_command(t, page, FLASH_CMD_WP)) {
+			return -1;
+		}
+		/* it should be done after one page, but in case it isn't */
+		len -= SAM4L_PAGE_SIZE;
+		addr += SAM4L_PAGE_SIZE;
+	}
+	return 0;
+}
+
+/*
+ * Erase flash across the addresses specified by addr and len
+ */
+static int
+sam4l_flash_erase(struct target_flash *f, target_addr addr, size_t len)
+{
+	target *t = f->t;
+	uint16_t page;
+
+	DEBUG("SAM4L: flash erase address 0x%08x for %d bytes\n", 
+		(unsigned int) addr, (unsigned int) len);
+	/*
+	 *  NB: if addr isn't aligned to a page boundary, or length
```

Comment:
You're safe here if you put the right values in the memory map.  GDB will do the right thing.  The comment is fine though.